### DISCUSSION TITLE
Constraining SimulationBody movement on path

### DISCUSSION THREAD

**Discussion by NordelDev** [dyn4j/dyn4j#171] [General]
Hi William,

I found the following [example](https://github.com/dyn4j/dyn4j-samples/blob/master/src/main/java/org/dyn4j/samples/LinkTerrain.java), where you define a path/terrain at which a `SimulationBody` slides at connected `Link`s.

Is it currently possible to define a constraint, so that a `SimulationBody` can only move along a defined path?

I thought about the following scenarios:

![grafik](https://user-images.githubusercontent.com/34217758/106125288-41937380-615c-11eb-87fd-a4c690858f25.png)

![grafik](https://user-images.githubusercontent.com/34217758/106125298-4526fa80-615c-11eb-9639-e11497083e45.png)

Bascially `SimulationBody` objects are pushed from one side inside the "chain" and move according to applied forces towards the end, where they then stop.

I think the best real world example would be:
![grafik](https://user-images.githubusercontent.com/34217758/106125486-86b7a580-615c-11eb-8dce-315c04fd912b.png)

The path would be an invisible objects which doesn't collide with the `SimulationBody`s. (this I know how to achieve)

The best case would be, that the approach can cover both scenarios
*  curved path (picture 1)
*  rectangular path (picture 2)

Do you have any ideas how this constraining could be achieved?

Thank you in advance for your suggestions :)

Carsten

**Comment by wnbittle** [comment]
It's achievable, but not easy.  There's two approaches that I'm thinking about - but I think it depends on what control you want.

One, which is the easiest, is that you control the body with another (invisible) body via the MotorJoint and you translate the controller body (the invisible one) through the defined path.  If you were looking for something more complex, there's a lot of literature on predefined movement out there.

The second way would be to define a new unary joint (one body) that takes as input an axis of travel and anchor point.  You evolve the axis over time based on the current tangent of the curve.  The unary joint is pretty straight forward, but the evolving part is what's hard.  It boils down to the closest point on a curve to a point.  A good example of the problem this has is https://bl.ocks.org/mbostock/8027637.  Move the mouse where the point is equal distance from two sections of the curve - you can see that it can quickly "jump" between points on the graph.  We don't know where it's supposed to be because the body is technically moving freely until the joint clamps it - so if the body was moving sufficiently fast, it could quickly jump to a new position on the curve.

**Reply by NordelDev** [reply]
Thank you very much for your reply.

I guess for my intended goal the second approach would be the more suitable one.
The link you provided was the missing link. I also had an idea of an additional `SimulationBody` which I translate along the path and restrict the other `SimulationBody` with a temporary / contantly updating `PrismaticJoint`.

I will try my luck the next days and will keep you updated :)

//Edit: For anyone also interested in this topic

I found the following thread for Box2D, which I will try to recreate.
[How to set a box2d body to follow a strict path](https://gamedev.stackexchange.com/questions/162163/how-to-set-a-box2d-body-to-follow-a-strict-path)
